### PR TITLE
chore: add reporting of nightly results to test rail

### PR DESCRIPTION
**Pull request by StateOf-Vlado**
Sending test results in TestRail for nightly regression added:
TESTRAIL_RUN_ID renamed to TESTRAIL_RUN_NAME.
TESTRAIL_RUN_NAME responsible for the report in TestRail:
- if empty - report will be not sent
- if contains value:
   - if in TestRail we already have TestRun with the same name, the report will be sent to it
   - else TestRun will be created with this name and test cases that were  started in the current session

So for release testing, we can create TestRun with some name and all test cases (or what we need to test for the release), paste the name in the TESTRAIL_RUN_NAME field, and start tests. 

For Nightly regression job:
The scheduled test run will be started with an empty TESTRAIL_RUN_NAME field. In the testrail fixture, I'm using BUILD_URL (Jenkins) to add it to TestRail, and also I'm using it in testrail configs, to generate a name for TestRun. So if TESTRAIL_RUN_NAME is empty and the current run is nightly, TESTRAIL_RUN_NAME will be f'Nightly regression {datetime.now():%d.%m.%Y}'

Also in the current PR, I added rerun tests if it failed. 
pytest-rerunfailures==11.1.2
By default, the test will be executed once more time.

Test Run: https://ci.status.im/job/status-desktop/job/e2e/job/Test_nightly/12/ (After testing, the Test_nightly will be deleted)
TestReil: https://ethstatus.testrail.net/index.php?/runs/view/13461&group_by=cases:title&group_order=asc

### CODE REVIEW

**Review by Valentina1133 — APPROVED** [review]

**Comment by anastasiyaig** [comment]
why it is Retest and why it is posted 3 times?

Screenshot 2023-10-30 at 12 49 15

File: ci/Jenkinsfile [hunk 1]
```diff
@@ -70,6 +70,7 @@ pipeline {
     LD_LIBRARY_PATH = "${SQUISH_DIR}/lib:${SQUISH_DIR}/python3/lib:${LD_LIBRARY_PATH}"
 
     TESTRAIL_URL = 'https://ethstatus.testrail.net'
+    TESTRAIL_PROJECT_ID = 17
```
Comment (anastasiyaig) [inline ci/Jenkinsfile:73]: any specific reason we keep the hardcoded value? Can we make a constant please?

**Review by anastasiyaig — CHANGES_REQUESTED** [review]

**Comment by anastasiyaig** [comment]
also, one more point: please exclude the skipped tests from being re-launched.

image

**Comment by glitchminer** [comment]
@StateOf-Vlado , I think we probably need to clarify what the "retry" status actually means in the report. Just now it seems to be used for "skipped" tests. I'm expecting something like this:

- passed: assertion passed
- failed: assertion failed
- retry: multiple retries required to pass (this one might be optional, perhaps it would make it easy to find flaky tests but they could also be considered passed)
- skipped: intentionally skipped
- error: fail unrelated to assertion

**Comment by StateOf-Vlado** [comment]
> why it is Retest and why it is posted 3 times? @StateOf-Vlado
> 
> Screenshot 2023-10-30 at 12 49 15
> Expected behavior:
> 
> 1. if tests are passed in allure report - the test rail run should also be green
> 2. if something requires to retest (like what?) , please post a clear reason why
> 3. please provide an example when a test is failed (u can break any test on purpose, just dont forget to bring it back later)

1. Give me permission in TestRail to set up test case statuses. I would change this status to UNTESTED or SKIPPED

File: ci/Jenkinsfile (hunk 1 above)
Comment (StateOf-Vlado) [reply ci/Jenkinsfile:73]: Could you please give me example, of how can I improve it?

Comment (anastasiyaig) [reply ci/Jenkinsfile:73]: in my initial question i asked if we can keep a constant for this somewhere. if not then thats it @StateOf-Vlado

**Comment by StateOf-Vlado** [comment]
> also, one more point: please exclude the skipped tests from being re-launched.
> 
> image
Where did you get it?
I can't find it here https://ethstatus.testrail.net/index.php?/tests/view/3264224&group_by=cases:title&group_order=asc&group_id=-1

**Comment by StateOf-Vlado** [comment]
> @StateOf-Vlado , I think we probably need to clarify what the "retry" status actually means in the report. Just now it seems to be used for "skipped" tests. I'm expecting something like this:
> 
> * passed: assertion passed
> * failed: assertion failed
> * retry: multiple retries required to pass (this one might be optional, perhaps it would make it easy to find flaky tests but they could also be considered passed)
> * skipped: intentionally skipped
> * error: fail unrelated to assertion

I'm expecting something like this:
FAILED - test Faild
PASS - tests Pass
SKIPPED - test skipped
UNTESTED - not executed
TESTING -  execution test 

Give me permission to setup statuses in TestRail

**Comment by anastasiyaig** [comment]
@StateOf-Vlado to clarify - permissions should be granted to your personal user or automation user the report is generated by?

**Comment by StateOf-Vlado** [comment]
> @StateOf-Vlado to clarify - permissions should be granted to your personal user or automation user the report is generated by?

to my account vlado@status.im

**Comment by StateOf-Vlado** [comment]
PyTest doesn't allow marking tests as SKIPPED because it skips tests before all fixtures.
So before executing tests I create a Test Run with all test cases in session and mark them as SKIPPED.
During the test execution, I marked tests as Pass or Failed
For the Pass result, I did not add any comment
For Failed results, I added a stack trace in the comment

The last test run:
- TestRail: https://ethstatus.testrail.net/index.php?/runs/view/13480&group_by=cases:title&group_order=asc
-Jenkins: https://ci.status.im/job/status-desktop/job/e2e/job/Test_nightly/14/

**Comment by anastasiyaig** [comment]
@StateOf-Vlado perfect, great job, thank you very much , i love it!

2 more cents from me:
1. when test failed, we post a log to the comment in results which is good! However , the log is super huge, maybe there is a way to grab the last 10 lines where actual error is posted only?
2. when test is skipped, is there any way to post the reason of it to be skipped to the results (reason value from pytest.mark.skip)?

**Comment by StateOf-Vlado** [comment]
> @StateOf-Vlado perfect, great job, thank you very much , i love it!
> 
> 2 more cents from me:
> 
> 1. when test failed, we post a log to the comment in results which is good! However , the log is super huge, maybe there is a way to grab the last 10 lines where actual error is posted only?
> 2. when test is skipped, is there any way to post the reason of it to be skipped to the results (reason value from pytest.mark.skip)?

1. The log collected by pytest
2 Fixed

**Review by anastasiyaig — APPROVED** [review]

**Pull request merged**